File: o9pkg.go
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/knusbaum/go9p"
	"github.com/knusbaum/go9p/fs"
	"github.com/knusbaum/go9p/proto"
)

/* 
 * o9pkg.go - High-speed Go-based 9P Installer 
 * Dynamically serves files from disk to prevent sync issues.
 */

var instScript = `#!/bin/rc
echo 'Installing o9 Toolchain (Go-powered)...'
cp /n/o9/bin/o9c /bin/o9c
cp /n/o9/include/o9.h /sys/include/o9.h
cp /n/o9/lib/libo9.a /sys/lib/libo9.a
chmod +x /bin/o9c
echo 'o9 Installation Complete.'
`

type DiskFile struct {
	*fs.BaseFile
	path string
}

func (f *DiskFile) Read(fid *fs.FFid, offset uint64, count uint32) ([]byte, error) {
	data, err := os.ReadFile(f.path)
	if err != nil {
		return nil, err
	}
	if offset >= uint64(len(data)) {
		return nil, nil
	}
	end := offset + uint64(count)
	if end > uint64(len(data)) {
		end = uint64(len(data))
	}
	return data[offset:end], nil
}

func main() {
	user := os.Getenv("USER")
	if user == "" {
		user = "scott"
	}

	fsys, root := fs.NewFS("o9", user, 0755)
	
	root.AddChild(fs.NewStaticFile(fsys.NewStat("install", user, user, 0444), []byte(instScript)))
	
	binDir := fs.NewStaticDir(fsys.NewStat("bin", user, user, 0755|proto.DMDIR))
	root.AddChild(binDir)
	
	incDir := fs.NewStaticDir(fsys.NewStat("include", user, user, 0755|proto.DMDIR))
	root.AddChild(incDir)
	
	libDir := fs.NewStaticDir(fsys.NewStat("lib", user, user, 0755|proto.DMDIR))
	root.AddChild(libDir)

	srcDir := fs.NewStaticDir(fsys.NewStat("src", user, user, 0755|proto.DMDIR))
	root.AddChild(srcDir)

	cmdDir := fs.NewStaticDir(fsys.NewStat("cmd", user, user, 0755|proto.DMDIR))
	srcDir.AddChild(cmdDir)

	o9cSrcDir := fs.NewStaticDir(fsys.NewStat("o9c", user, user, 0755|proto.DMDIR))
	cmdDir.AddChild(o9cSrcDir)

	// Dynamically serve files from disk
	addDiskFile(fsys, binDir, "o9c", "o9c/o9c", user)
	addDiskFile(fsys, incDir, "o9.h", "o9.h", user)
	addDiskFile(fsys, libDir, "libo9.a", "libo9.a", user)
	addDiskFile(fsys, o9cSrcDir, "o9.y", "o9c/o9.y", user)

	fmt.Println("o9 Go-Powered Installer listening on :9009 (Dynamic Sync)")
	log.Fatal(go9p.Serve("0.0.0.0:9009", fsys.Server()))
}

func addDiskFile(fsys *fs.FS, parent *fs.StaticDir, name, path, user string) {
	df := &DiskFile{
		path: path,
	}
	df.BaseFile = fs.NewBaseFile(fsys.NewStat(name, user, user, 0444))
	parent.AddChild(df)
}

File: o9pkg.c
#include <u.h>
#include <libc.h>
#include <thread.h>
#include <fcall.h>
#include <9p.h>

/* 
 * o9pkg.c - The o9 "Mountable Installer" Fileserver
 */

static char *inst_script = 
    "#!/bin/rc\n"
    "echo 'Installing o9 Toolchain...'\n"
    "cp /n/o9/bin/o9c /bin/o9c\n"
    "cp /n/o9/include/o9.h /sys/include/o9.h\n"
    "cp /n/o9/lib/libo9.a /sys/lib/libo9.a\n"
    "chmod +x /bin/o9c\n"
    "echo 'o9 Installation Complete.'\n";

static void
serve_host_file(Req *r, char *path)
{
    int fd;
    long n;
    char *buf;

    fd = p9open(path, OREAD);
    if(fd < 0){
        respond(r, "could not open host file");
        return;
    }
    
    buf = emalloc9p(r->ifcall.count);
    n = pread(fd, buf, r->ifcall.count, r->ifcall.offset);
    close(fd);

    if(n < 0){
        free(buf);
        respond(r, "read error");
        return;
    }

    r->ofcall.data = buf;
    r->ofcall.count = n;
    respond(r, nil);
    free(buf);
}

static void
fsread(Req *r)
{
    char *name = r->fid->file->dir.name;

    if(strcmp(name, "install") == 0){
        readstr(r, inst_script);
        respond(r, nil);
        return;
    }
    
    if(strcmp(name, "o9c") == 0) { serve_host_file(r, "o9c/o9c"); return; }
    if(strcmp(name, "o9.h") == 0) { serve_host_file(r, "o9.h"); return; }
    if(strcmp(name, "libo9.a") == 0) { serve_host_file(r, "libo9.a"); return; }

    respond(r, "not found");
}

Srv o9pkg_srv = {
    .read = fsread,
};

void
usage(void)
{
    fprint(2, "usage: o9pkg [-a address]\n");
    threadexitsall("usage");
}

void
threadmain(int argc, char **argv)
{
    Tree *t;

    ARGBEGIN{
    case 'a':
        EARGF(usage());
        break;
    }ARGEND

    t = alloctree(nil, nil, 0555, nil);
    o9pkg_srv.tree = t;

    createfile(t->root, "install", nil, 0444, nil);
    File *bin = createfile(t->root, "bin", nil, DMDIR|0555, nil);
    createfile(bin, "o9c", nil, 0555, nil);
    File *inc = createfile(t->root, "include", nil, DMDIR|0555, nil);
    createfile(inc, "o9.h", nil, 0444, nil);
    File *lib = createfile(t->root, "lib", nil, DMDIR|0555, nil);
    createfile(lib, "libo9.a", nil, 0444, nil);

    /* Post the service locally */
    threadpostmountsrv(&o9pkg_srv, "o9pkg", nil, MREPL);
    
    print("o9 Installer Fileserver Online.\n");
    print("\nTo serve on network (Bash-safe):\n");
    print("aux/listen -t 'tcp!*!9009' /bin/9pserve -u \"$NAMESPACE/o9pkg\"\n\n");
    
    /* Allow the process to continue in the background */
    threadmaybackground();
    
    /* Keep the main thread alive but dormant */
    for(;;)
        sleep(10000);
}
